File: android/app/src/main/java/dev/surge/worktick/WTSettings.kt
package dev.surge.worktick

import android.content.Context
import androidx.core.content.edit

/**
 * User-configurable settings backed by SharedPreferences. Read by the Settings
 * UI, the poller, and the widget provider.
 */
object WTSettings {
    private const val PREFS = "worktick_settings"
    private const val KEY_RATE = "hourly_rate"
    private const val KEY_TITLE = "event_title"
    private const val KEY_LAST_POLL = "last_poll_at"
    private const val KEY_LAST_ERROR = "last_poll_error"
    private const val KEY_PORTAL_URL = "portal_url"
    private const val KEY_NOTIFY_LEAD_MIN = "notify_lead_min"

    private const val DEFAULT_TITLE = "McCrary Summer Work"
    private const val DEFAULT_RATE = 30.0
    private const val DEFAULT_LEAD_MIN = 5

    fun hourlyRate(context: Context): Double =
        prefs(context).getFloat(KEY_RATE, DEFAULT_RATE.toFloat()).toDouble()

    fun setHourlyRate(context: Context, value: Double) {
        prefs(context).edit { putFloat(KEY_RATE, value.toFloat()) }
    }

    fun eventTitle(context: Context): String =
        prefs(context).getString(KEY_TITLE, DEFAULT_TITLE) ?: DEFAULT_TITLE

    fun setEventTitle(context: Context, value: String) {
        prefs(context).edit { putString(KEY_TITLE, value.trim()) }
    }

    fun lastPollAt(context: Context): Long = prefs(context).getLong(KEY_LAST_POLL, 0L)
    fun setLastPollAt(context: Context, value: Long) {
        prefs(context).edit { putLong(KEY_LAST_POLL, value) }
    }

    fun lastPollError(context: Context): String? = prefs(context).getString(KEY_LAST_ERROR, null)
    fun setLastPollError(context: Context, value: String?) {
        prefs(context).edit {
            if (value == null) remove(KEY_LAST_ERROR) else putString(KEY_LAST_ERROR, value)
        }
    }

    /** Portal URL opened when the user taps a clock-in/out reminder notification. */
    fun portalUrl(context: Context): String =
        prefs(context).getString(KEY_PORTAL_URL, "")?.trim().orEmpty()

    fun setPortalUrl(context: Context, value: String) {
        prefs(context).edit { putString(KEY_PORTAL_URL, value.trim()) }
    }

    /** Minutes before a block boundary to fire the clock-in/out reminder. 0 = at the boundary. */
    fun notifyLeadMinutes(context: Context): Int =
        prefs(context).getInt(KEY_NOTIFY_LEAD_MIN, DEFAULT_LEAD_MIN).coerceAtLeast(0)

    fun setNotifyLeadMinutes(context: Context, value: Int) {
        prefs(context).edit { putInt(KEY_NOTIFY_LEAD_MIN, value.coerceAtLeast(0)) }
    }

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}

File: ios/Platform/WTSettings.swift
// WTSettings.swift
// User-configurable settings, persisted to a UserDefaults suite shared with the
// widget extension via App Group `group.dev.surge.worktick`. Mirrors the
// Android WTSettings.kt key set so behavior matches across platforms.
//
// IMPORTANT: For the App Group suite to work, both the host app target AND the
// widget extension target must have the App Groups capability with
// `group.dev.surge.worktick` enabled. See docs/IOS.md.

import Foundation

public enum WTSettingsKey {
    public static let appGroup = "group.dev.surge.worktick"

    public static let hourlyRate          = "hourly_rate"
    public static let eventTitle          = "event_title"
    public static let selectedCalendarIDs = "selected_calendar_ids"
    public static let lastPollAt          = "last_poll_at"          // unix seconds
    public static let lastPollError       = "last_poll_error"
    public static let eventKitAuthorized  = "event_kit_authorized"  // cached
    public static let liveActivityEnabled = "live_activity_enabled" // user opt-in
    public static let portalURL           = "portal_url"
    public static let notifyLeadMinutes   = "notify_lead_min"
}

public enum WTSettingsDefaults {
    public static let hourlyRate: Double = 30.0
    public static let eventTitle: String = "McCrary Summer Work"
    public static let notifyLeadMinutes: Int = 5
}

public final class WTSettings {
    public static let shared = WTSettings()

    private let store: UserDefaults

    public init(suiteName: String = WTSettingsKey.appGroup) {
        // UserDefaults(suiteName:) returns nil only for the reserved
        // "globalDomain" / app-bundle suite names; the App Group ID is always
        // safe but if it's mis-configured (no entitlement), reads/writes go
        // nowhere silently. We fall back to .standard so the app still works
        // in local dev without an App Group set up.
        self.store = UserDefaults(suiteName: suiteName) ?? .standard
    }

    // MARK: Hourly rate

    public var hourlyRate: Double {
        get {
            let v = store.double(forKey: WTSettingsKey.hourlyRate)
            return v == 0 ? WTSettingsDefaults.hourlyRate : v
        }
        set { store.set(newValue, forKey: WTSettingsKey.hourlyRate) }
    }

    // MARK: Event title

    public var eventTitle: String {
        get { store.string(forKey: WTSettingsKey.eventTitle) ?? WTSettingsDefaults.eventTitle }
        set {
            let trimmed = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            store.set(trimmed.isEmpty ? WTSettingsDefaults.eventTitle : trimmed,
                      forKey: WTSettingsKey.eventTitle)
        }
    }

    // MARK: Selected calendars

    /// Empty array means "use all available calendars" (default behavior on
    /// first launch, matches Android which polls every visible calendar).
    public var selectedCalendarIDs: [String] {
        get { store.stringArray(forKey: WTSettingsKey.selectedCalendarIDs) ?? [] }
        set { store.set(newValue, forKey: WTSettingsKey.selectedCalendarIDs) }
    }

    // MARK: Last poll bookkeeping

    public var lastPollAt: Date? {
        get {
            let v = store.double(forKey: WTSettingsKey.lastPollAt)
            return v == 0 ? nil : Date(timeIntervalSince1970: v)
        }
        set {
            if let d = newValue {
                store.set(d.timeIntervalSince1970, forKey: WTSettingsKey.lastPollAt)
            } else {
                store.removeObject(forKey: WTSettingsKey.lastPollAt)
            }
        }
    }

    public var lastPollError: String? {
        get { store.string(forKey: WTSettingsKey.lastPollError) }
        set {
            if let s = newValue {
                store.set(s, forKey: WTSettingsKey.lastPollError)
            } else {
                store.removeObject(forKey: WTSettingsKey.lastPollError)
            }
        }
    }

    // MARK: EventKit auth cache

    /// Cached auth state. Truth comes from EKEventStore.authorizationStatus
    /// at launch — this is just a hint for widget timeline scheduling.
    public var eventKitAuthorized: Bool {
        get { store.bool(forKey: WTSettingsKey.eventKitAuthorized) }
        set { store.set(newValue, forKey: WTSettingsKey.eventKitAuthorized) }
    }

    // MARK: Live Activity opt-in

    /// User opt-in for Live Activity (lock screen + Dynamic Island). Off by
    /// default — the home screen widget is the primary surface, the Live
    /// Activity is optional. Defaults to false; UserDefaults.bool returns
    /// false for missing keys, which is exactly what we want.
    public var liveActivityEnabled: Bool {
        get { store.bool(forKey: WTSettingsKey.liveActivityEnabled) }
        set { store.set(newValue, forKey: WTSettingsKey.liveActivityEnabled) }
    }

    // MARK: Portal URL

    /// URL opened when the user taps a clock-in / clock-out reminder
    /// notification. Empty string disables the link (notification still fires).
    public var portalURL: String {
        get { store.string(forKey: WTSettingsKey.portalURL) ?? "" }
        set {
            let trimmed = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            store.set(trimmed, forKey: WTSettingsKey.portalURL)
        }
    }

    // MARK: Notify lead time

    /// Minutes before a block start/end to fire a reminder notification.
    /// 0 = right at the boundary.
    public var notifyLeadMinutes: Int {
        get {
            let raw = store.object(forKey: WTSettingsKey.notifyLeadMinutes) as? Int
            return max(0, raw ?? WTSettingsDefaults.notifyLeadMinutes)
        }
        set { store.set(max(0, newValue), forKey: WTSettingsKey.notifyLeadMinutes) }
    }
}
